packages/ampd-sdk/src/grpc/connection_pool.rs

```diff
@@ -0,0 +1,394 @@
+#[derive(Debug, Clone)]
+pub enum ConnectionState {
+    /// The connection is established and ready for service calls.
+    /// Contains the active gRPC transport channel.
+    Connected(transport::Channel),
+
+    /// The connection is not available and no reconnection attempt is in progress.
+    /// This is the initial state when the connection pool starts, or the final state after all reconnection attempts have failed.  
+    Disconnected,
+
+    /// A reconnection attempt is currently in progress.
+    /// The connection pool is actively trying to re-establish the connection after a failure.
+    /// Clients should wait for the next state change rather than immediately failing their requests.
+    Reconnecting,
+}
+
+impl PartialEq for ConnectionState {
+    fn eq(&self, other: &Self) -> bool {
+        discriminant(self) == discriminant(other)
+    }
+}
+
+/// A handle that provides access to connection state changes and allows sending
+/// messages to the connection pool.
+///
+/// This struct is used by clients to:
+/// - Monitor connection state changes via the `connection_receiver`
+/// - Send messages (like connection failure reports) to the connection pool
+/// - Coordinate with the connection pool for automatic reconnection
+#[derive(Clone, Debug)]
+pub struct ConnectionHandle {
+    connection_receiver: watch::Receiver<ConnectionState>,
+    message_sender: mpsc::Sender<ClientMessage>,
+}
+
+impl ConnectionHandle {
+    /// This method encapsulates all the connection state management logic:
+    /// - If connected, returns the channel immediately
+    /// - If disconnected, inform clients
+    /// - If reconnecting, waits for the connection to complete and returns the channel
+    ///
+    /// Clients should use this method instead of manually handling connection states.
+    pub async fn get_connected_channel(&mut self) -> Result<transport::Channel, Error> {
+        timeout(DEFAULT_SERVER_TIMEOUT, async {
+            loop {
+                let connection_state = self.connection_receiver.borrow_and_update().clone();
+
+                match connection_state {
+                    ConnectionState::Connected(channel) => {
+                        return Ok(channel);
+                    }
+                    ConnectionState::Disconnected => {
+                        self.send_client_message(ClientMessage::ConnectionFailed(
+                            "client requested connection while disconnected".to_string(),
+                        ))
+                        .await?;
+
+                        let mut receiver = self.connection_receiver.clone();
+                        receiver.changed().await.into_report()?;
+                    }
+                    ConnectionState::Reconnecting => {
+                        let mut receiver = self.connection_receiver.clone();
+                        receiver.changed().await.into_report()?;
+                    }
+                }
+            }
```

Comment:
can we simplify to this? (pseudo code)
```
loop{
    let state = receiver.borrow_and_update();
    match state{
       ConnectionState::Connected(channel) => return Connected<channel>,
       ConnectionState::Shutdown/Disconnected => return Shutdown
       _ =>     receiver.changed().await.into_report()?;
    }    
}
```
We will need another enum for the return from the handle to tell the client to disconnect when the pool shuts down (I don't want to shut down with errors)